# KNN (K-Nearest-Neighbors)
# KNN（K-最近的鄰）

KNN is a simple concept: define some distance metric between the items in your dataset, and find the K closest items. You can then use those items to predict some property of a test item, by having them somehow "vote" on it.

As an example, let's look at the MovieLens data. We'll try to guess the rating of a movie by looking at the 10 movies that are closest to it in terms of genres and popularity.

To start, we'll load up every rating in the data set into a Pandas DataFrame:

KNN 是一個簡單的概念：定義數資料集中的項目之間的一些距離指標，並查找 K 最接近的項目。然後，您可以使用這些項目來預測測試的某些屬性，讓他們以某種方式"投票"。

例如，讓我們看看電影的數據。嘗試查看 10 部電影在流行和受歡迎程度方面最接近它來猜測電影的評級。

首先，我們將把資料集中的每一個評級載入：

In [1]:
import pandas as pd

r_cols = ['user_id', 'movie_id', 'rating']
ratings = pd.read_csv('./ml-100k/u.data', sep='\t', names=r_cols, usecols=range(3))
ratings.head()


user_id  movie_id  rating
0        0        50       5
1        0       172       5
2        0       133       1
3      196       242       3
4      186       302       3

Now, we'll group everything by movie ID, and compute the total number of ratings (each movie's popularity) and the average rating for every movie:

現在，我們將按電影ID對所有內容進行分組，並計算每部電影的收視率總數（每部電影的受歡迎程度）和平均評分：

In [2]:
import numpy as np

movieProperties = ratings.groupby('movie_id').agg({'rating': [np.size, np.mean]})  # movie_id為 1 的電影，有432個評分。
movieProperties.head()

rating          
           size      mean
movie_id                 
1           452  3.878319
2           131  3.206107
3            90  3.033333
4           209  3.550239
5            86  3.302326

The raw number of ratings isn't very useful for computing distances between movies, so we'll create a new DataFrame that contains the normalized number of ratings. So, a value of 0 means nobody rated it, and a value of 1 will mean it's the most popular movie there is.

原始的評分數量對於計算電影之間的距離不是很有用，因此我們將創建一個新的 DataFrame，其中包含標準化的評分數量。因此，0 值意味著沒有人對它進行評級，而 1 值將意味著它是最受歡迎的電影。 (每部電影流行的程度在0~1之間。)

為了進行正規化，將所有電影的最多評分數量(np.max)與最少評分數量(np.min)進行比較。

In [3]:
movieNumRatings = pd.DataFrame(movieProperties['rating']['size'])
movieNormalizedNumRatings = movieNumRatings.apply(lambda x: (x - np.min(x)) / (np.max(x) - np.min(x)))
movieNormalizedNumRatings.head()

size
movie_id          
1         0.773585
2         0.222985
3         0.152659
4         0.356775
5         0.145798

Now, let's get the genre information from the u.item file. The way this works is there are 19 fields, each corresponding to a specific genre - a value of '0' means it is not in that genre, and '1' means it is in that genre. A movie may have more than one genre associated with it.

While we're at it, we'll put together everything into one big Python dictionary called movieDict. Each entry will contain the movie name, list of genre values, the normalized popularity score, and the average rating for each movie:

現在，讓我們載入美國電影資料，這部作品的方式是有19個類別，每個領域都對應於一個特定的類別 - '0'值意味著它不屬於該類別，而'1'意味著它屬於該類別。電影可能不止一種類別與它有關。

當我們在做的時候，我們會把所有東西都整理成一本叫做電影《Toy Story (1995)》的Python dist。每個項目將包含電影名稱、類別清單、人氣評分和每部電影的平均評分：

In [4]:
movieDict = {}
with open('./ml-100k/u.item', encoding='unicode_escape') as f:
    temp = ''
    for line in f:
#         line.decode("ISO-8859-1")
        fields = line.rstrip('\n').split('|')
        movieID = int(fields[0])
        name = fields[1]
        genres = fields[5:25]
        genres = map(int, genres)
        movieDict[movieID] = (name, np.array(list(genres)), movieNormalizedNumRatings.loc[movieID].get('size'), movieProperties.loc[movieID].rating.get('mean'))


For example, here's the record we end up with for movie ID 1, "Toy Story":

In [5]:
print(movieDict[1])

('Toy Story (1995)', array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), 0.7735849056603774, 3.8783185840707963)


Now let's define a function that computes the "distance" between two movies based on how similar their genres are, and how similar their popularity is. Just to make sure it works, we'll compute the distance between movie ID's 2 and 4:

現在，讓我們根據兩部電影的類別相似程度以及它們的受歡迎程度來定義一個函數來計算它們之間的"距離"。為了確保它的工作原理，我們將計算電影ID的2和4之間的距離：

In [6]:
from scipy import spatial

def ComputeDistance(a, b): #ComputeDistance()計算距離
    genresA = a[1]
    genresB = b[1]
    genreDistance = spatial.distance.cosine(genresA, genresB)
    popularityA = a[2]
    popularityB = b[2]
    popularityDistance = abs(popularityA - popularityB)
    return genreDistance + popularityDistance
    
ComputeDistance(movieDict[2], movieDict[4])



0.8004574042309892

Remember the higher the distance, the less similar the movies are. Let's check what movies 2 and 4 actually are - and confirm they're not really all that similar:

記住距離越遠，電影就越不相似。讓我們檢查什麼是電影 2 和 4 實際上 - 並確認他們不是真的那麼相似：

In [7]:
print(movieDict[2])
print(movieDict[4])


('GoldenEye (1995)', array([0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0]), 0.22298456260720412, 3.2061068702290076)
('Get Shorty (1995)', array([0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), 0.3567753001715266, 3.550239234449761)


Now, we just need a little code to compute the distance between some given test movie (Toy Story, in this example) and all of the movies in our data set. When the sort those by distance, and print out the K nearest neighbors:

現在，我們只需要一個小代碼來計算一些給定的測試電影（玩具總動員，在本例中）和我們數據集中的所有電影之間的距離。當按距離排序這些，並列印出K最近的鄰居：

In [8]:
import operator

def getNeighbors(movieID, K):
    distances = []
    for movie in movieDict:
        if (movie != movieID):
            dist = ComputeDistance(movieDict[movieID], movieDict[movie])
            distances.append((movie, dist))
    distances.sort(key=operator.itemgetter(1))
    neighbors = []
    for x in range(K):
        neighbors.append(distances[x][0])
    return neighbors

K = 10
avgRating = 0
neighbors = getNeighbors(1, K)
for neighbor in neighbors:
    avgRating += movieDict[neighbor][3]
    print (movieDict[neighbor][0] + " " + str(movieDict[neighbor][3]))
    
avgRating /= K

Liar Liar (1997) 3.156701030927835
Aladdin (1992) 3.8127853881278537
Willy Wonka and the Chocolate Factory (1971) 3.6319018404907975
Monty Python and the Holy Grail (1974) 4.0664556962025316
Full Monty, The (1997) 3.926984126984127
George of the Jungle (1997) 2.685185185185185
Beavis and Butt-head Do America (1996) 2.7884615384615383
Birdcage, The (1996) 3.4436860068259385
Home Alone (1990) 3.0875912408759123
Aladdin and the King of Thieves (1996) 2.8461538461538463


While we were at it, we computed the average rating of the 10 nearest neighbors to Toy Story:

計算10個鄰近《玩具總動員》的平均評分：

In [9]:
avgRating

3.3445905900235564

How does this compare to Toy Story's actual average rating?

《玩具總動員》實際平均評分

In [11]:
movieDict[1]

('Toy Story (1995)',
 array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
 0.7735849056603774,
 3.8783185840707963)

Not too bad!

預測評分是3.34分，實際是3.87分，還不算太糟。

## Activity

Our choice of 10 for K was arbitrary - what effect do different K values have on the results?

Our distance metric was also somewhat arbitrary - we just took the cosine distance between the genres and added it to the difference between the normalized popularity scores. Can you improve on that?

我們選擇 10 代表 K 是任意的 - 不同的 K 值對結果有什麼影響？

我們的距離指標也有些武斷 - 我們只是採取了類別之間的距離， 並把它添加到人氣分數之間的差異。你能改進一下嗎？

In [20]:
import operator

def getNeighbors(movieID, K):
    distances = []
    for movie in movieDict:
        if (movie != movieID):
            dist = ComputeDistance(movieDict[movieID], movieDict[movie])
            distances.append((movie, dist))
    distances.sort(key=operator.itemgetter(1))
    neighbors = []
    for x in range(K):
        neighbors.append(distances[x][0])
    return neighbors

K = 5
avgRating = 0
neighbors = getNeighbors(1, K)
for neighbor in neighbors:
    avgRating += movieDict[neighbor][3]
    print (movieDict[neighbor][0] + " " + str(movieDict[neighbor][3]))
    
avgRating /= K

Liar Liar (1997) 3.156701030927835
Aladdin (1992) 3.8127853881278537
Willy Wonka and the Chocolate Factory (1971) 3.6319018404907975
Monty Python and the Holy Grail (1974) 4.0664556962025316
Full Monty, The (1997) 3.926984126984127


In [21]:
avgRating

3.7189656165466287